Use either a browser or sailfish webview for logging in.

b100dian opened the pull request:
Added an auth-server that actually listens on port 3000 and if you do connect with the browser it displays "You can now switch back to Sailslack".

~~Moved the sailfish webview to a lazy-loaded component as it still crashes if not run from terminal \o/~~
Crash is worked around, webview is still lazy loaded but now we can use it in WebViewPage which makes for better virtual keyboard support.
The sailfish webview redirect actually works, I left the comment for reference from my initial impressions..

Here's the stack of the crash.. https://github.com/sailfishos/sailfish-components-webview/issues/26#issuecomment-753697401
(The link actually has the fix for the crash now!)

To be able to build this, you need to tick the webview in Qt's Sailfish IDE Tools -> Options -> Sailfish OS -> Build Engine -> Manage Build Targets -> search for "webview" and tick sailfish-components-webview-qt5
![image](https://user-images.githubusercontent.com/71930/103492650-f627c700-4e34-11eb-876c-b8cd13d32d99.png)

b100dian committed:
Try to use sailfish webview for loging in. Works if you start harbour-sailslack from ssh.

b100dian committed:
Slightly better urlChanged listener

b100dian committed:
Offer choice of browser vs webview login (latter not working yet)

b100dian committed:
Lazy load webview in case the user uses the terminal. Stop the auth server if so.

b100dian committed:
revert version change

b100dian committed:
Merge branch 'master' into sailfish-webview

b100dian committed:
Fix webview crash and enabled back WebViewPage for better virtual keyboard support

b100dian committed:
Copy link option + auto offer browser login

danvratil commented on src/authserver.cpp:
```diff
@@ -0,0 +1,54 @@
+    int verbEnd = result.indexOf(' ') + 1;
+    int httpStart = result.lastIndexOf(" HTTP/");
+    auto urlWithoutHost = result.mid(verbEnd, httpStart - verbEnd);
+    emit resultUrlAvailable(QString("http://localhost:") + QString::number(server.serverPort()) + urlWithoutHost);
```
Use `QStringLiteral("http://localhost:%1%2").arg(server.serverPort()).arg(urlWithoutHost)` to build the string.

danvratil commented on src/authserver.cpp:
```diff
@@ -0,0 +1,54 @@
+    QString result(resultBytes);
+
+    // Replace "GET " with http://localhost:3000 and " HTTP 1.x" with nothing
+    int verbEnd = result.indexOf(' ') + 1;
+    int httpStart = result.lastIndexOf(" HTTP/");
+    auto urlWithoutHost = result.mid(verbEnd, httpStart - verbEnd);
```
Make everything `const`

danvratil commented on src/authserver.cpp:
```diff
@@ -0,0 +1,54 @@
+    auto socket = server.nextPendingConnection();
+    Q_ASSERT(socket);
+    socket->waitForReadyRead();
+    QByteArray resultBytes = socket->readLine();
```
`const QByteArray`

danvratil commented on src/authserver.cpp:
```diff
@@ -0,0 +1,54 @@
+
+bool AuthServer::listen(quint16 port) {
+    Q_ASSERT(!server.isListening());
+    bool result = server.listen(QHostAddress::LocalHost, port);
```
`const bool`

danvratil commented on src/authserver.h:
```diff
@@ -0,0 +1,32 @@
+private Q_SLOTS:
+    void clientConnected();
+
+signals:
```
`Q_SIGNALS:` (for consistency with `Q_SLOTS` above)

b100dian replied:
Oh thanks, I wasn't sure about the equivalency of these two syntax.

danvratil commented on qml/pages/LoginPage.qml:
```diff
@@ -1,17 +1,17 @@
     id: page
 
     property string processId: Math.random().toString(36).substring(7)
     property string startUrl: "https://slack.com/oauth/authorize?scope=client&client_id=" + slackClientId + "&redirect_uri=https%3A%2F%2Flocalhost%3A3000%2Foauth%2Fcallback"
```
Looks like there's typo in this line - the return `redirect_uri` must be `http`, not `https` - not sure how this ever worked :)

b100dian replied:
Ooops, you're right. Looks like one of the previous commits did this.. It may be working since the redirect URL was specified by you when creating the Slack client id.

danvratil commented on qml/pages/LoginPage.qml:
```diff
@@ -21,42 +21,117 @@ Page {
+                onClicked: {
+                    // Closing the 3000 port as even though the webview would successfully connect to it, the timing it navigates away from this page makes
+                    // the webview crash (at onResultUrlAvailable signal).
+                    server.close();
```
Without the server running, the login via webview never finishes for me.

To avoid the crash, you should close the client socket after sending the response HTML in AuthServer, then the webview no longer crashes, because the connection is terminated before the webview is destroyed by the pagestack pop.

b100dian replied:
I'll investigate this more. The first commit of this branch there wass no problem with the webivew connecting locally (it was actually the only way to get back the token)

b100dian replied:
I think I got this: the webview was a bit more picky about my "http" response, and it was still 'loading' when destroyed. Adding Content-Length to the response seems to have fixed that. Let me know if this works on your end.

danvratil commented on qml/pages/LoginPage.qml:
```diff
@@ -21,42 +21,117 @@ Page {
+        switch (status) {
+        case PageStatus.Active:
+            server.listen(3000);
+            Qt.openUrlExternally(page.startUrl + "&state=" + page.processId);
```
This causes the web browser to be automatically opened when the page is first shown. Is that intended as default behaviour, where the alternative options are just for "advanced" users? Or would it be better to let the user to choose the right method for themselves?

b100dian replied:
What would be your preference?
I actually have more than one browser so this pops a dialog to let me chose the browser. But now I realize that fresh installs will directly start the login sequence in the browser (while I am still able to dismiss that dialog).
This was a last minute 'great idea' that maybe this new screen is not user friendly. I am open to suggestions if this should be enabled or removed.

danvratil replied:
Hmm I would say let's keep the screen (for debugging purposes if nothing else) and go to webview by default (feels like a better flow). Regarding the browser, have you tried whether this works when the user users an Android browser, like Firefox?

b100dian replied:
No, I havent yet tried with an AD browser (can you register one as url handler?)
I will try to show the webview first and offer an "Advanced" pull-down menu

b100dian replied:
Done, see latest commit @danvratil

danvratil reviewed (changes requested):
Looks really good, thanks a lot for this patch! There are some minor nitpicks, small issue, nothing major.

b100dian committed:
Added content length to HTTP response + comments addressed

danvratil commented on src/authserver.cpp:
```diff
@@ -0,0 +1,57 @@
+    const QString replyHTML = "<head><title>Sailslack connected</title><meta name='viewport' content='width=device-width, initial-scale=1'/> </head><body><header>You can now switch back to Sailslack</header></body>";
+
+    socket->write(QStringLiteral("HTTP/1.0 200 OK\r\nContent-type:text/html\r\nContent-Length:%1\r\n\r\n%2").arg(replyHTML.length()).arg(replyHTML).toUtf8());
+    socket->flush();
```
I still think you should close the connection here - in a standard HTTP communication, the server sends response and closes the connection, so I think we should do it the same way. `Content-Length` works of course, but it's just a workaround :)

b100dian replied:
doh.. that was probably it. Will try next, thanks!

b100dian committed:
Webview visible first with Advanced login PullDownMenu

Also closed the socket after sending the HTTP response.
The webview now has a header because I noticed that swipong may
be disabled before a connection is made and this would mean losing
the ability to access Advanced pulley in the future / in some cases.

danvratil reviewed (approved):
Thanks a lot for this change! Great job!
